### PR TITLE
Feature to show hidden card count

### PR DESCRIPTION
**Pull request by Rajat-Dabade**
This PR will add a feature to show the hidden card count with the label `Cards Hidden`

### CODE REVIEW

File: webapp/src/store/cards.ts
Comment (sbishel) [inline webapp/src/store/cards.ts:326]: This should be named `GetHiddenCards`

File: webapp/i18n/en.json
Comment (sbishel) [inline webapp/i18n/en.json:260]: Are you sure this needs localized?

File: webapp/src/components/gallery/gallery.test.tsx
Comment (sbishel) [inline webapp/src/components/gallery/gallery.test.tsx:180]: Add a snapshot check here as well.

File: webapp/src/components/gallery/gallery.tsx
Comment (sbishel) [inline webapp/src/components/gallery/gallery.tsx:105]: Move to a component, then just call component where desired. Encapsulates the limit card message feature

File: webapp/src/components/kanban/__snapshots__/kanban.test.tsx.snap
Comment (sbishel) [inline webapp/src/components/kanban/__snapshots__/kanban.test.tsx.snap:282]: Based on the code changes, nothing should change since your checking 
`if (limitedCard.length > 0) {`
and there are no changes to test data.
In addition, this item is `<span>0</span` which shouldn't happen.

File: webapp/src/components/gallery/gallery.tsx
Comment (sbishel) [inline webapp/src/components/gallery/gallery.tsx:34]: nit - `limited` sounds like a `boolean` variable. Why not name this `hiddenCards`?

File: webapp/src/components/kanban/kanban.tsx
Comment (sbishel) [inline webapp/src/components/kanban/kanban.tsx:58]: I wouldn't confuse "LimitedCards" with "HiddenGroups". I would keep those things separate. If you create a "LimitedCards" component, you can simply add that compoent after the "hiddenGroups". see below, Line 331. None of this code would be necessary.

File: webapp/src/components/kanban/kanban.test.tsx
Comment (sbishel) [inline webapp/src/components/kanban/kanban.test.tsx:83]: Add a test for displaying the component.

File: webapp/src/components/kanban/kanbanHiddenColumnItem.test.tsx
Comment (sbishel) [inline webapp/src/components/kanban/kanbanHiddenColumnItem.test.tsx:119]: Nice, add snapshot check as well.

File: webapp/src/components/kanban/kanbanHiddenColumnItem.tsx
Comment (sbishel) [inline webapp/src/components/kanban/kanbanHiddenColumnItem.tsx:58]: This code shouldn't be needed if you separate LimitedCards and HiddenGroups

File: webapp/src/components/table/table.test.tsx
Comment (sbishel) [inline webapp/src/components/table/table.test.tsx:205]: Check snapshot as well.

File: webapp/src/components/kanban/kanbanHiddenColumnItem.tsx
Comment (sbishel) [inline webapp/src/components/kanban/kanbanHiddenColumnItem.tsx:67]: This is why your snapshot is changing...

**Review by sbishel — CHANGES_REQUESTED** [review]
@Rajat-Dabade You should create a `<HiddenCardCount>` component. Then in the various views, you can simply call that component. This will encapsulate that functionality. Also, we should check the cloud license first, so we don't do anything else if there are no limits. An on-premise server should not ever need to call `getHiddenCards`. 

Let me know if any of my comments need further explanation. Thanks

File: webapp/src/components/kanban/__snapshots__/kanban.test.tsx.snap
Comment (sbishel) [reply webapp/src/components/kanban/__snapshots__/kanban.test.tsx.snap:282]: Realize now, the 0 is because its a "Hidden Column" not "limited-card-count"

**Comment by sbishel** [comment]
@Rajat-Dabade - Could you also add screen shots to the PR. Thanks.

File: webapp/src/blocks/card.ts
Comment (Pinjasaur) [inline webapp/src/blocks/card.ts:15]: question: could this just be added to the Block type? similar to in my PR: https://github.com/mattermost/focalboard/blob/a863e43fb0876d21d7996d92f5c3452a99c25437/webapp/src/blocks/block.ts#L44

File: webapp/src/blocks/card.ts
Comment (Pinjasaur) [inline webapp/src/blocks/card.ts:40]: I don't think we want to default it to false, right? It should just be passed through? https://github.com/mattermost/focalboard/blob/a863e43fb0876d21d7996d92f5c3452a99c25437/webapp/src/blocks/block.ts#L63

similar to my comment above, could this be moved to `createBlock` instead of being in `createCard`?

File: webapp/src/test/testBlockFactory.ts
Comment (Pinjasaur) [inline webapp/src/test/testBlockFactory.ts:122]: Same question as my other comments, can it just be passed through?

File: webapp/src/blocks/card.ts
Comment (Rajat-Dabade) [reply webapp/src/blocks/card.ts:15]: Yes, nice catch. I will update it.

File: webapp/src/blocks/card.ts
Comment (Rajat-Dabade) [reply webapp/src/blocks/card.ts:40]: Yes. Fixed

File: webapp/src/test/testBlockFactory.ts
Comment (Rajat-Dabade) [reply webapp/src/test/testBlockFactory.ts:122]: Done

**Comment by Rajat-Dabade** [comment]
Here are the screenshot, how what hidden cards look like:

For Status view:

Screenshot 2022-05-23 at 5 47 55 PM
Screenshot 2022-05-23 at 5 48 08 PM

For List view: 
Screenshot 2022-05-23 at 5 48 38 PM

For Gallery view:
Screenshot 2022-05-23 at 5 49 14 PM

File: webapp/src/store/cards.ts
Comment (mgdelacroix) [inline webapp/src/store/cards.ts:326]: nit: to keep the naming and casing consistent, I'd suggest something in plural and a bit more explicit, derived from `getCurrentBoardCards` as it's actually getting the inforrmation from that selector:

```suggestion
export const getCurrentBoardHiddenCards = createSelector(
```

File: mac/Focalboard/whatsnew.txt
Comment (mgdelacroix) [inline mac/Focalboard/whatsnew.txt:1]: Don't think this file should be updated as part of this PR

File: webapp/src/components/gallery/gallery.tsx
Comment (mgdelacroix) [inline webapp/src/components/gallery/gallery.tsx:33]: Reviewing the information flow for the `gallery`, `kanban` and `table` components, the `board`, `cards` and other pieces of the store are being fetched in the `CenterContent` component, so I think the `hiddenCards` should be fetched there as well, and passed through

**Review by mgdelacroix — COMMENTED** [review]
Looking really good!

File: webapp/src/components/kanban/kanbanHiddenColumnItem.tsx
Comment (Rajat-Dabade) [reply webapp/src/components/kanban/kanbanHiddenColumnItem.tsx:61]: Fixed

File: webapp/src/store/cards.ts
Comment (Rajat-Dabade) [reply webapp/src/store/cards.ts:326]: Sure, will change it.

File: webapp/src/components/gallery/gallery.tsx
Comment (Rajat-Dabade) [reply webapp/src/components/gallery/gallery.tsx:33]: True, I will make the changes. Thanks

File: mac/Focalboard/whatsnew.txt
Comment (mgdelacroix) [inline mac/Focalboard/whatsnew.txt:5]: nit: minor change left behind

**Review by mgdelacroix — APPROVED** [review]
Thanks for applying the changes @Rajat-Dabade! LGTM! :rocket:

File: mac/Focalboard/whatsnew.txt
Comment (Rajat-Dabade) [reply mac/Focalboard/whatsnew.txt]: Fixed. Thanks

File: webapp/src/components/workspace.tsx
Comment (sbishel) [inline webapp/src/components/workspace.tsx:39]: We should wrap this with the normal Cloud checks. `utils.IsCloudLicense() and GetCloudLimits`

This will slow down on-premise and paid accounts. by looping through cards unnecessarily.

File: webapp/src/components/hiddenCardCount/hiddenCardCount.tsx
Comment (sbishel) [inline webapp/src/components/hiddenCardCount/hiddenCardCount.tsx:13]: Didn't notice before, but you are only using the `.length` of array. Probably better to just get and pass an `Integer` rather than a `Card[]`.

**Review by sbishel — CHANGES_REQUESTED** [review]
I'm sorry for not noticing this before, but since you moved the retrieval of hidden cards into `workspace.tsx`, I noticed that you don't need the HiddenCards (Cards[]) at all. You only use the Len of that array. So, I would suggest changing the `HIddenCards` property from a Card[] to `HIddenCardsCount` as an Int. Defaulting the property to 0, will allow you to not need to change the test code as well.

Let me know if I missed where the actually cards are needed. Also, we should be aware of performance implications for non-limited installations and wrap those with checks.

File: webapp/src/components/hiddenCardCount/hiddenCardCount.tsx
Comment (Pinjasaur) [inline webapp/src/components/hiddenCardCount/hiddenCardCount.tsx:20]: looks like you'll want to run `npm run i18n-extract` to update the `en.json` file

I'd recommend updating your branch with `release-7.0` before you do that as my PR that was merged yesterday updated the `npm run i18n-extract` command

**Review by Pinjasaur — APPROVED** [review]
looks good @Rajat-Dabade! small note re: updating the i18n file(s) and it looks like there's a couple merge conflicts, but otherwise LGTM after addressing that

File: webapp/src/components/hiddenCardCount/hiddenCardCount.tsx
Comment (Rajat-Dabade) [reply webapp/src/components/hiddenCardCount/hiddenCardCount.tsx:20]: Okay, got it.

**Comment by Rajat-Dabade** [comment]
@sbishel, Just had a word with @jespino, he will be taking care of wrapping the necessary code with the `isCloudLicense` whenever it is needed and optimising the selector code. Thank you.

**Review by sbishel — APPROVED** [review]

**Pull request merged**
